### DISCUSSION TITLE
Change Kingfisher default cache expiration globally

### DISCUSSION THREAD

**Discussion by SandraNPR** [onevcat/Kingfisher#2087] [Q&A] [answered]
It was observed that iOS devices which had an old logo cached were not properly expiring the cache lifetime set in Akamai (24 hours). The default settings for diskCacheAccessExtendingExpiration is .cacheTime, meaning every time the logo is accessed from the cache, its lifetime is extended by 1 week.

We should switch the behavior to diskCacheAccessExtendingExpiration(.none) and verify this on the server for cache lifetime. 
I'm trying to do this globally and not per image because we have Kingfisher used in a lot of different places for images. 

I attempted to change the default options by setting the diskCacheAccessExtendingExpiration = .none
When I tested this, it did not take effect. Maybe something is wrong with how I'm trying to change it, there's probably something I'm missing here. An example on how to do this globally would be great.

```swift
@objc func setKingfisherCacheExpiration() {
       let currentCacheDefaultOptions = KingfisherManager.shared.defaultOptions
       var cache = KingfisherParsedOptionsInfo(currentCacheDefaultOptions)
       cache.diskCacheAccessExtendingExpiration = .none
```

**Comment by onevcat** [comment]
Not sure how did you set it globally, but can you try this?

```swift
KingfisherManager.shared.defaultOptions
            .append(.diskCacheAccessExtendingExpiration(.none))
```

**Reply by SandraNPR** [reply]
Thanks for the response! 😃 I changed the function but it did not work.
```
    @objc func setKingfisherCacheExpiration() {
        var currentCacheDefaultOptions = KingfisherManager.shared.defaultOptions
        currentCacheDefaultOptions.append(.diskCacheAccessExtendingExpiration(.none))        
    }
```
    

I called the function in our - AppDelegate in the didFinishLaunchingWithOptions
```

 -(BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions {
    [DDLog addLogger:[DDTTYLogger sharedInstance]];
```

Not sure about the best way I could be setting this globally. With the change you suggested, the output was still .cacheTime for diskCacheAccessExtendingExpiration

`  - loadDiskFileSynchronously : false
  ▿ diskStoreWriteOptions : NSDataWritingOptions
    - rawValue : 0
  - memoryCacheExpiration : nil
  - memoryCacheAccessExtendingExpiration : Kingfisher.ExpirationExtending.cacheTime
  - diskCacheExpiration : nil
  - diskCacheAccessExtendingExpiration : Kingfisher.ExpirationExtending.cacheTime
  - processingQueue : nil
  - progressiveJPEG : nil
  - alternativeSources : nil
  - retryStrategy : nil
  - lowDataModeSource : nil
  - onDataReceived : nil`

**Comment by onevcat** [comment, marked as answer]
Instead of 

```swift
var currentCacheDefaultOptions = KingfisherManager.shared.defaultOptions
currentCacheDefaultOptions.append(.diskCacheAccessExtendingExpiration(.none))    
```

You need

```swift
KingfisherManager.shared.defaultOptions
            .append(.diskCacheAccessExtendingExpiration(.none))
```

Or 

```swift
var currentCacheDefaultOptions = KingfisherManager.shared.defaultOptions
currentCacheDefaultOptions.append(.diskCacheAccessExtendingExpiration(.none))    
KingfisherManager.shared.defaultOptions = currentCacheDefaultOptions
```

It is a value type, not a reference type.

**Reply by SandraNPR** [reply]
Thanks for the help! This works 🥳
